src/solver.rs

```diff
@@ -293,54 +293,44 @@ pub trait DependencyProvider {
     /// assign [`String`] as placeholder.
     type M: Eq + Clone + Debug + Display;
 
-    /// [Decision making](https://github.com/dart-lang/pub/blob/master/doc/solver.md#decision-making)
-    /// is the process of choosing the next package
-    /// and version that will be appended to the partial solution.
+    /// The type returned from `prioritize`. The resolver does not care what type this is
+    /// as long as it can pick a largest one and clone it.
     ///
-    /// Every time such a decision must be made, the resolver looks at all the potential valid
-    /// packages that have changed, and a asks the dependency provider how important each one is.
-    /// For each one it calls `prioritize` with the name of the package and the current set of
-    /// acceptable versions.
-    /// The resolver will then pick the package with the highes priority from all the potential valid
-    /// packages.
+    /// [`Reverse`](std::cmp::Reverse) can be useful if you want to pick the package with
+    /// the fewest versions that match the outstanding constraint.
+    type Priority: Ord + Clone;
+
+    /// The kind of error returned from these methods.
     ///
-    /// The strategy employed to prioritize packages
-    /// cannot change the existence of a solution or not,
-    /// but can drastically change the performances of the solver,
-    /// or the properties of the solution.
-    /// The documentation of Pub (PubGrub implementation for the dart programming language)
-    /// states the following:
+    /// Returning this signals that resolution should fail with this error.
+    type Err: Error + 'static;
+
+    /// Determine the order in which versions are chosen for packages.
     ///
-    /// > Pub chooses the latest matching version of the package
-    /// > with the fewest versions that match the outstanding constraint.
-    /// > This tends to find conflicts earlier if any exist,
-    /// > since these packages will run out of versions to try more quickly.
-    /// > But there's likely room for improvement in these heuristics.
+    /// Decisions are always made for the highest priority package first. The order of decisions
+    /// determines which solution is chosen and can drastically change the performances of the
+    /// solver. If there is a conflict between two package versions, the package with the higher
+    /// priority is preserved and the lower priority gets discarded. Usually, you want to decide
```

Comment:
> If there is a conflict between two package versions, the priority is preserved and the lower priority gets discarded.

I'm not following. What scenario are you trying to describe?

Reply:
Added the word "decision".

Say:
choose A
decide A 2
choose B 
decide B 2

and then we discover A 2 -> C 1, B 2 -> C 2.

In this case, we will next try B 1, not A 1: The version of A 2 gets preserved, while B 2 gets discarded.

Reply:
Perhaps: "If there is a conflict between two package versions, decisions will be backtracked until the lower priority package version is discarded preserving the higher priority package." or something.